core/slice.go

```diff
@@ -0,0 +1,484 @@
+		}
+	}
+	if len(blocks) > 0 {
+		sort.Slice(blocks, func(i, j int) bool {
+			return blocks[i].NumberU64() < blocks[j].NumberU64()
+		})
+		// Insert one by one as chain insertion needs contiguous ancestry between blocks
+		for i := range blocks {
+			sl.Append(blocks[i])
+		}
+	}
+}
+
+// addFutureBlock checks if the block is within the max allowed window to get
+// accepted for future processing, and returns an error if the block is too far
+// ahead and was not added.
+func (sl *Slice) addFutureBlock(block *types.Block) error {
+	max := uint64(time.Now().Unix() + maxTimeFutureBlocks)
+	if block.Time() > max {
+		return fmt.Errorf("future block timestamp %v > allowed %v", block.Time(), max)
+	}
+	if !sl.futureBlocks.Contains(block.Hash()) {
+		sl.futureBlocks.Add(block.Hash(), block)
+	}
+	return nil
+}
+
+func (sl *Slice) update() {
+	futureTimer := time.NewTicker(1 * time.Second)
+	defer futureTimer.Stop()
+	defer sl.wg.Done()
+	for {
+		select {
+		case <-futureTimer.C:
+			sl.procFutureBlocks()
+		case <-sl.quit:
+			return
+		}
+	}
+}
+
+// CalcTd calculates the TD of the given header using PCRC and CalcHLCRNetDifficulty.
+func (sl *Slice) CalcTd(header *types.Header) ([]*big.Int, error) {
+	// Iterate ancestors, stopping when a TD value is found in cache or a coincident block is found.
+	// If coincident is found, ask dom client for TD at that block
+	aggDiff := new(big.Int)
+	cursor := header
+	for {
+		// First, check if this block's TD is already cached
+		td := sl.hc.GetTd(cursor.Hash(), (*cursor.Number[types.QuaiNetworkContext]).Uint64())
+		if td != nil {
+			// Add the difficulty we accumulated up till this block
+			blockTd := td[types.QuaiNetworkContext]
+			td[types.QuaiNetworkContext] = blockTd.Add(blockTd, aggDiff)
+			return td, nil
+		}
+
+		// If not cached, check if this block coincides with a dominant chain
+		order, err := sl.GetDifficultyOrder(cursor)
+		if err != nil {
+			return nil, err
+		} else if order < types.QuaiNetworkContext {
+			// TODO: Ask dom to CalcTd on coincident block
+			td, err = sl.domClient.CalcTd(context.Background(), header)
+			err = errors.New("TODO: Ask dom to CalcTd")
+			if err != nil {
+				return nil, err
+			} else {
+				blockTd := td[types.QuaiNetworkContext]
+				td[types.QuaiNetworkContext] = blockTd.Add(blockTd, aggDiff)
+			}
+		}
+
+		// If not cached AND not coincident, aggregate the difficulty and iterate to the parent
+		aggDiff = aggDiff.Add(aggDiff, cursor.Difficulty[types.QuaiNetworkContext])
+		parentHash := cursor.ParentHash[types.QuaiNetworkContext]
+		cursor = sl.hc.GetHeader(cursor.Hash(), (*cursor.Number[types.QuaiNetworkContext]).Uint64())
+		if cursor == nil {
+			return nil, fmt.Errorf("Unable to find parent: %s", parentHash)
+		}
+	}
+}
+
+var big2e256 = new(big.Int).Exp(big.NewInt(2), big.NewInt(256), big.NewInt(0)) // 2^256
+
+// This function determines the difficulty order of a block
+func (sl *Slice) GetDifficultyOrder(header *types.Header) (int, error) {
```

Comment:
I thought we decided this does not work. We need to ask dom if it was accepted as a higher order block.